components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java

```diff
@@ -580,6 +614,266 @@ protected void initiateAuthenticationRequest(HttpServletRequest request, HttpSer
+                        "Cannot retrieve the query parameters.");
+            }
+            return null;
+        }
+
+        String scopeString = getQueryParameter(context, OIDCAuthenticatorConstants.FEDERATED_TOKEN_SCOPE);
+        if (StringUtils.isBlank(scopeString)) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("No query parameter " + OIDCAuthenticatorConstants.FEDERATED_TOKEN_SCOPE +
+                        " found in federated token sharing, IDP: " + authenticatorName);
+            }
+            return null;
+        }
+        /*
+        The requested scopes for particular authenticator should come with the authenticator name separated
+        by a semicolon.
+        i.e. A valid string:
+        When Google Calender has read write scopes and Microsoft Authenticator has https://googleapis.calender scope
+        A valid requested scopes string:
+        federated_token_scope=Google Calander;read write,Microsoft Authenticator;https://googleapis.calender
+         */
+        if (!scopeString.contains(SEMI_COLON_DELIMITER)) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Query parameter " + OIDCAuthenticatorConstants.FEDERATED_TOKEN_SCOPE + " is missing " +
+                        SEMI_COLON_DELIMITER + " delimiter in federated token sharing, IDP: " +
+                        authenticatorName);
+            }
+            return null;
+        }
+
+        String[] scopeSegments = StringUtils.split(scopeString, COMMA_DELIMITER);
+        StringBuilder filteredScopes = new StringBuilder();
+
+        for (String scopesFollowedByAuthenticator : scopeSegments) {
+            String[] scopes = StringUtils.split(scopesFollowedByAuthenticator, SEMI_COLON_DELIMITER);
+            if (ArrayUtils.getLength(scopes) == 2 && authenticatorName.equals(scopes[0])) {
+                filteredScopes.append(scopes[1]).append(SPACE);
+            }
+        }
+
+        String requestedScopes = filteredScopes.toString();
+        if (LOG.isDebugEnabled() && StringUtils.isBlank(requestedScopes)) {
+            LOG.debug("No valid values found for the authenticator " + authenticatorName + " in the query parameter " +
+                    OIDCAuthenticatorConstants.FEDERATED_TOKEN_SCOPE + " for federated token sharing, IDP: " +
+                    authenticatorName);
+        }
+
+        return filteredScopes.toString();
+    }
+
+    /**
+     * This method evaluates whether application has requested to share the token. The first priority is given to the
+     * authenticator parameters set at the adaptive script. Then the query parameters.
+     *
+     * @param context The authentication context.
+     * @return Whether the application has requested to share the token.
+     */
+    private boolean requestedToShareFederatedToken(AuthenticationContext context) {
+
+        String authenticatorName = getFederatedAuthenticatorName(context);
```

Comment:
this is only used in the debug logs, lets use this directly.

Reply:
Fixed d2b40f9255b4a5ef3b8f9d44bd58ccdba9f50f70